engine/apps/public_api/views/shift_swap.py

```diff
@@ -0,0 +1,98 @@
+from rest_framework.response import Response
+from rest_framework.serializers import BaseSerializer
+
+from apps.api.permissions import AuthenticatedRequest
+from apps.api.views.shift_swap import BaseShiftSwapViewSet
+from apps.auth_token.auth import ApiTokenAuthentication
+from apps.public_api.throttlers.user_throttle import UserThrottle
+from apps.schedules.models import ShiftSwapRequest
+from apps.user_management.models import User
+from common.api_helpers.custom_fields import TimeZoneAwareDatetimeField
+from common.api_helpers.exceptions import BadRequest
+from common.api_helpers.mixins import RateLimitHeadersMixin
+
+logger = logging.getLogger(__name__)
+
+
+class ShiftSwapViewSet(RateLimitHeadersMixin, BaseShiftSwapViewSet):
+    # set authentication and permission classes
+    authentication_classes = (ApiTokenAuthentication,)
+    permission_classes = (IsAuthenticated,)
+
+    # public API customizations
+    throttle_classes = [UserThrottle]
+
+    def get_queryset(self):
+        schedule_id = self.request.query_params.get("schedule_id", None)
+        beneficiary = self.request.query_params.get("beneficiary", None)
+        benefactor = self.request.query_params.get("benefactor", None)
+        starting_after = self.request.query_params.get("starting_after", None)
+        untaken = self.request.query_params.get("untaken", "false") == "true"
+
+        if starting_after:
+            f = TimeZoneAwareDatetimeField()
+            # trigger datetime format validation
+            # will raise ValidationError if invalid timestamp is provided
+            starting_after = f.to_internal_value(starting_after)
+        else:
+            starting_after = timezone.now()
+
+        # base queryset filters by organization
+        queryset = super().get_queryset()
+        queryset = queryset.filter(swap_start__gte=starting_after)
+
+        if schedule_id:
+            queryset = queryset.filter(schedule__public_primary_key=schedule_id)
+
+        if beneficiary:
+            queryset = queryset.filter(beneficiary__public_primary_key=beneficiary)
+
+        if benefactor:
+            queryset = queryset.filter(benefactor__public_primary_key=benefactor)
+
+        if benefactor:
+            queryset = queryset.filter(benefactor__public_primary_key=benefactor)
+
+        if untaken:
+            queryset = queryset.filter(benefactor__isnull=True)
+
+        return queryset.order_by("swap_start")
+
+    def get_object(self):
+        public_primary_key = self.kwargs["pk"]
+        try:
+            return self.get_queryset().get(public_primary_key=public_primary_key)
+        except ShiftSwapRequest.DoesNotExist:
+            raise NotFound
+
+    def _get_user(self, field_name: str):
+        """Return user from ID given by field_name.
+
+        Fallback to request.user if no ID is provided.
+        """
+        user_pk = self.request.data.pop(field_name, None)
+        if not user_pk:
+            user = self.request.user
```

Comment:
`self.request.user` is it the user that created an API token? If so, maybe it makes sense to make the `beneficiary` and `benefactor` fields required, without any fallbacks?
I think we don't show API key creators anywhere, so having a fallback could introduce extra confusion.

<img width="1409" alt="Screenshot 2023-08-11 at 16 18 48" src="https://github.com/grafana/oncall/assets/20116910/76ed7c39-f682-41ad-93b4-08b3988deded">

Reply:
Makes sense :+1: I guess it will be good to discuss end-user/roles API tokens at some point.